x/stake/keeper.go

```diff
@@ -33,319 +29,356 @@ func NewKeeper(cdc *wire.Codec, key sdk.StoreKey, ck bank.Keeper, codespace sdk.
+//
+// The correct subset is retrieved by iterating through an index of the
+// validators sorted by power, stored using the ValidatorsByPowerKey. Simultaniously
+// the current validator records are updated in store with the
+// ValidatorsBondedKey. This store is used to determine if a validator is a
+// validator without needing to iterate over the subspace as we do in
+// GetValidators.
+//
+// Optionally also return the validator from a retrieve address if the validator has been bonded
+func (k Keeper) updateBondedValidators(ctx sdk.Context, store sdk.KVStore, pool Pool,
+	OptionalRetrieve sdk.Address) (retrieveBonded bool, retrieve Validator) {
+
+	// clear the current validators store, add to the ToKickOut temp store
+	toKickOut := make(map[string][]byte) // map[key]value
+	iterator := store.SubspaceIterator(ValidatorsBondedKey)
 	for ; iterator.Valid(); iterator.Next() {
-		key := iterator.Key()
-		addr := AddrFromKey(key)
 
-		// get the zero abci validator from the ToKickOut iterator value
 		bz := iterator.Value()
 		var validator Validator
-		err := k.cdc.UnmarshalBinary(bz, &validator)
-		if err != nil {
-			panic(err)
-		}
-		bz, err = k.cdc.MarshalBinary(validator.abciValidatorZero(k.cdc))
-		if err != nil {
-			panic(err)
-		}
+		k.cdc.MustUnmarshalBinary(bz, &validator)
 
-		store.Set(GetAccUpdateValidatorKey(addr), bz)
-		store.Delete(key)
+		addr := validator.Address
+
+		// iterator.Value is the validator object
+		toKickOut[string(addr)] = iterator.Value()
+		store.Delete(iterator.Key())
 	}
 	iterator.Close()
 
-	return validators[:i] // trim
-}
-
-// TODO this is madly inefficient because need to call every time we set a candidate
-// Should use something better than an iterator maybe?
-// Used to determine if something has just been added to the actual validator set
-func (k Keeper) isNewValidator(ctx sdk.Context, store sdk.KVStore, address sdk.Address) bool {
 	// add the actual validator power sorted store
-	maxVal := k.GetParams(ctx).MaxValidators
-	iterator := store.ReverseIterator(subspace(ValidatorsKey)) // largest to smallest
-	for i := 0; ; i++ {
-		if !iterator.Valid() || i > int(maxVal-1) {
+	maxValidators := k.GetParams(ctx).MaxValidators
+	iterator = store.ReverseSubspaceIterator(ValidatorsByPowerKey) // largest to smallest
+	i := 0
+	for ; ; i++ {
+		if !iterator.Valid() || i > int(maxValidators-1) {
 			iterator.Close()
 			break
 		}
 		bz := iterator.Value()
-		var val Validator
-		err := k.cdc.UnmarshalBinary(bz, &val)
-		if err != nil {
-			panic(err)
-		}
-		if bytes.Equal(val.Address, address) {
-			return true
+		var validator Validator
+		k.cdc.MustUnmarshalBinary(bz, &validator)
+
+		_, found := toKickOut[string(validator.Address)]
+		if found {
+
+			// remove from ToKickOut group
+			delete(toKickOut, string(validator.Address))
+		} else {
+
+			// if it wasn't in the toKickOut group it means
+			// this wasn't a previously a validator, therefor
+			// update the validator/to reflect this
```

Comment:
Sanity check `validator.Status != sdk.Bonded` ?

Reply:
added this sanity check in bondValidator